### PR TITLE
Add watch indicators

### LINKED ISSUE #1001: Fade watched videos instance-wide
Really useful and unhealthy feature for those of us with compulsive disorders

### LINKED ISSUE #1195: Add grey filter over thumbnails of watched videos, do not hide from feed.
Please add grey filter over thumbnails of watched videos, I don't want them to be completely hidden from my feed.

_Originally posted by @utlasidyo in https://github.com/omarroth/invidious/issues/129#issuecomment-471329015_

### PR DESCRIPTION
**Pull request by WesVleuten**
Closes #1001
Closes #1195

image

With partial watch indicator if position is saved as shown in the image.

Curious to hear what you think!

### CODE REVIEW

**Comment by SamantazFox** [comment]
Oooh, very cool, indeed!
I've some things to say on code clarity, I'll try to write a review asap ^^

File: assets/js/watched_widget.js [hunk 1]
```diff
@@ -32,3 +32,29 @@ function mark_unwatched(target) {
...
+
+    if (percentage < 5) {
+        percentage = 5;
+    }
+    if (percentage > 90) {
+        percentage = 100;
+    }
+
+    indicator.style.width = percentage + '%';
+}
```
Comment (AHOHNMYC) [inline assets/js/watched_widget.js:60]:
```suggestion
document.querySelectorAll('.watched-indicator').forEach(function (indicator) {
    var watched_part = get_all_video_times()[indicator.dataset.id];
    var total = parseInt(indicator.dataset.length, 10);
    if (watched_part === undefined) {
        watched_part = total;
    }
    var percentage = Math.round((watched_part / total) * 100);

    if (percentage < 5) {
        percentage = 5;
    }
    if (percentage > 90) {
        percentage = 100;
    }

    indicator.style.width = percentage + '%';
});
```
I'd suggest using `forEach` construction, like everywhere in Invidious, and `.dataset` to accessing `data-something` values.
With `forEach` you can omit unneeded temporary variables.
Code is not tested, but have to work :)

File: src/invidious/views/components/item.ecr [hunk 2]
```diff
@@ -1,3 +1,5 @@
+<% item_watched = !item.is_a?(SearchChannel) && !item.is_a?(SearchPlaylist) && !item.is_a?(InvidiousPlaylist) && !item.is_a?(Category) && env.get("user") && env.get("user").as(User).watched && env.get("user").as(User).watched.index(item.id) != nil %>
```
[the remark this answers is not in the document]
Comment (SamantazFox) [reply src/invidious/views/components/item.ecr:1]: Similarly, you can chain the `env.get` part using the `.try?` method, like so:
```crystal
env.get?("user").try &.as(User).watched.index(item.id) != nil
```

`.get?` will return `nil` if "user" is not set in the HTTP context (`env`) storage. Then, if `nil` was returned, we stop here and return `nil`, orthewise we cast that returned value to a `User`-typed object and continue with the data we need.

**Review by SamantazFox — CHANGES_REQUESTED** [review]

**Comment by WesVleuten** [comment]
Thanks for the review, I'll look at the requested changes asap!

**Comment by SamantazFox** [comment]
Any updates on this?

**Comment by WesVleuten** [comment]
Requested changed have been made, and merge conflicts have been resolved

**Comment by SamantazFox** [comment]
Sorry for the lack of updates. I deployed this PR on the test instance (https://test.invidious.io).

I'm getting this error on various page, and the indicators doesn't show up:
image

The `watched_data` element mentioned is only present in the `history` and `subscriptions` feeds.
You should probably move your code to a separate JS file to avoid problems.

**Comment by SamantazFox** [comment]
Okay, so the indicators show up (Idk why my code on the test instance had issue compiling) but their width is always 100% (likely due to the error I mentioned above)

**Comment by WesVleuten** [comment]
It should be fixed now, I moved the indicator code to a different file and included it where needed

**Review by SamantazFox — APPROVED** [review]

**Comment by SamantazFox** [comment]
Thanks for contributing to invidious ^^

**Comment by WesVleuten** [comment]
Ofc! Happy to help!

**Pull request merged**